File: modules/cluster/jackknife.f90

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!
!
!   Module containing definitions and routines for manipulating data using the
!   jackknife re-sampling method.
!
!   Routines are:
!
!   _ cluster_init -> initialization of jackknife cluster;
!
!   _ clusterJK -> assigns the mean and the variance of the mean.
!
!   _ functionJK -> returns a clusterJK for a secondary r.v.
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

module jackknife
    use kinds, only: dp
    implicit none
    public

    ! definition of the jackknife cluster
    type JK
        real(dp), dimension(:), allocatable :: vec  ! sample vector
        integer :: D                        ! dimension of the sample
        real(dp) :: mean, var                   ! mean and variance of the mean
    end type JK

    ! interface for a real(dp) function of a real(dp) variable
!    interface
!        function func (x)
!            real(dp) :: x, func
!        end function
!    end interface
    
    ! interface for a JK cluster function
!    abstract interface
!        function jkf (f, X)
!            type (JK) :: jkf, X
!            procedure(func), pointer :: f
!            real(dp) :: temp
!            integer :: i
!        end function jkf
!    end interface    
    
contains

    ! Assignment of mean value and variance in a cluster structure
    subroutine JK_cluster (C)
        
        type(JK) :: C
        integer :: i,D
        
        D = C%D
        C%mean = 0
        do i=1, D, 1
            C%mean = C%mean + (C%vec(i))/(real(D, dp))
        end do
	
        do i=1, D, 1
            C%vec(i) = C%mean + (C%mean - C%vec(i))/(real(D-1, dp))
        end do
	
        C%var = 0
        do i=1, D, 1
            C%var = C%var + (C%vec(i) - C%mean)*(C%vec(i) - C%mean)
        end do
            C%var = C%var * (real(D-1, dp))/(real(D, dp))
    end subroutine JK_cluster


    subroutine JK_init (C, D)
        type(JK) :: C
        integer :: D
        
        C%D = D
        C%mean = 0
        C%var = 0
        allocate(C%vec(D))
    end subroutine JK_init


!    function JK_function (f, X) result(res)
        
        ! definition of arguments and result
!        type (JK) :: X, res
!        integer :: D
!        real(dp), external, pointer :: f
        ! ausiliary variables
!        integer :: i
!        real(dp) :: temp = 0
        
        ! initialization of the cluster for the result
!        D = X%D
!        call cluster_init(res,D);
        
        ! assignments
!        res%mean = f(X%mean);
!        do i=0, D-1, 1
!            res%vec(i) = f(X%vec(i));
!            temp = temp + (res%vec(i) - res%mean)*(res%vec(i) - res%mean);
!        end do
!        temp = temp * (real(dp)(D - 1)/real(dp)(D));
!        res%var = temp
        
!    end function JK_function

end module jackknife

File: modules/cluster/jackknife.c

/*******************************************************************************
 * 
 * 		File "jackknife.c"
 * 
 * File containing routines useful to handle cluster jackknife to compute
 * averages and variances of primary and secondary random variables
 * 
 * Routines are:
 * _ JKinit -> initialization of a cluster jackknife;
 * 
 * _ JKcluster -> assignment of mean and variance (of the mean) of the sample
 *		stored in the array contained in the cluster.
 *
 * _ JKfunction -> returns a cluster jackknife with samples, mean and variance
 *		(of the mean) of a secondary random variable defined as a function
 *		(passed as argument) of several primary random variables.
 *		For the sake of generality this routine requires a POINTER to an ARRAY
 *		of jackknife clusters as argument.
 * 
 ******************************************************************************/

#define JACKKNIFE_C

#include <stdio.h>
#include <stdlib.h>
#include "cluster.h"


/* Assignment of mean value and variance in a cluster structure */
void JKcluster (cluster *C)
{
	int i;
	int dim = C->Dim;
	
	C->Mean = 0;
	for(i=0; i<dim; i++)
		C->Mean += (C->Vec[i])/((double)dim);
	
	for(i=0; i<dim; i++)
		C->Vec[i] = C->Mean + (C->Mean - C->Vec[i])/((double)(dim - 1));
	
	C->Var = 0;
	for(i=0; i<dim; i++)
	{
		C->Var += (C->Vec[i] - C->Mean)*(C->Vec[i] - C->Mean);
		C->Var *= ((double)(dim - 1))/((double)dim);
	}
}


/* Cluster jackknive initialization */
void JKinit (cluster *C, int dim)
{
	C->Dim	= dim;
	C->Mean	= 0;
	C->Var= 0;
	C->Vec	= malloc(dim*sizeof(double));
}


/* Compute the JK of a secondary random variable */
cluster JKfunction (double (*f)(double *), int narg, cluster *X)
{
	int i, j;
	double temp = 0;
	double *args;

	cluster result;
	int dim = X->Dim;
	JKinit(&result,dim);

	args = malloc(narg*sizeof(double));
	for (j=0; j<narg; j++)
		args[j] = X[j].Mean;
	result.Mean = f(args);
	for(i=0; i<dim; i++) {
		for (j=0; j<narg; j++)
			args[j] = X[j].Vec[i];
		result.Vec[i] = f(args);
		temp += (result.Vec[i] - result.Mean)*(result.Vec[i] - result.Mean);
		temp *= ((double)(dim - 1)/(double)dim);
	}
	result.Var = temp;
	free(args);
	
	return result;
}
